hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java

```diff
@@ -142,51 +136,60 @@ public boolean containsValue(Object value) {
 
   @Override
   public R get(Object key) {
-    if(inMemoryMap.containsKey(key)) {
+    if (inMemoryMap.containsKey(key)) {
       return inMemoryMap.get(key);
-    } else if(diskBasedMap.containsKey(key)) {
+    } else if (diskBasedMap.containsKey(key)) {
       return diskBasedMap.get(key);
     }
     return null;
   }
 
   @Override
   public R put(T key, R value) {
-    try {
-      if (this.currentInMemoryMapSize < maxInMemorySizeInBytes || inMemoryMap.containsKey(key)) {
-        // Naive approach for now
-        if (estimatedPayloadSize == 0) {
-          this.estimatedPayloadSize = SpillableMapUtils.computePayloadSize(value, schema);
-          log.info("Estimated Payload size => " + estimatedPayloadSize);
-        }
-        if(!inMemoryMap.containsKey(key)) {
-          currentInMemoryMapSize += this.estimatedPayloadSize;
-        }
-        inMemoryMap.put(key, value);
-      } else {
-        diskBasedMap.put(key, value);
+    if (this.currentInMemoryMapSize < maxInMemorySizeInBytes || inMemoryMap.containsKey(key)) {
+      if (shouldEstimatePayloadSize && estimatedPayloadSize == 0) {
+        // At first, use the sizeEstimate of a record being inserted into the spillable map.
+        // Note, the converter may over estimate the size of a record in the JVM
+        this.estimatedPayloadSize =
+            keyConverter.sizeEstimate(key) + valueConverter.sizeEstimate(value);
+        log.info("Estimated Payload size => " + estimatedPayloadSize);
+      }
+      else if(shouldEstimatePayloadSize &&
```

Comment:
@vinothchandar Simple handling of sizes.

Reply:
how does estimating the size of entire map alleviate the issues you mentioned before?

Reply:
The inherent problem lies in the fact that an entry in the Map by itself is not a good gauge of how large the record is in the heap (as discussed before mainly due to shared objects). What essentially is required is to estimate the size of `a record` given `N records` collectively, that way the size of the shared object is `amortized` over N records and we get as close to the actual payload size as possible. Here, estimating the size of the map does that for us. It's still not dead accurate but comes very close, chose the number `100` because tried with a couple different payloads, nested and non-nested with shared objects and the size of the shared object get's amortized. We start with estimating the payload size for one record first, use that till we reach 100 records (since we don't want to OOM given a small memory setting and large record) and then update the payload estimate. We could choose a higher number say `1000` at which we want to re-estimate which might give use better amortization but the fear is that because of overestimation we may already start spilling to disk by then. 
The other option is to keep updating the estimate logarithmically, but my fear is for a huge HashMap, I'm not sure how the `ObjectSizeEstimator` performs in terms of CPU and Memory usage.

Reply:
lets do a follow on task to fix this more systematically.. if 100 works for now, may be we can keep it as it is. 

IIUC you are saying by estimating at a higher point in the object tree (graph), the sizing is more accurate? If so, then that makes sense..

Reply:
Yes, that's what I'm saying.